### DISCUSSION TITLE
Can we have an option to freeze/set dependency versions?

### DISCUSSION THREAD

**Discussion by pbaletkeman** [fastapi/fastapi#9947] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

# can something like this be implemented
app.configuration.requirements('pydantic',2.0.3)
```

### Description

Question:
Is there an option to freeze the version of any/all dependencies for FastAPI?
Recently, an update to Pydantic broke parts of FastAPI (https://github.com/tiangolo/fastapi/discussions/9942).
Luckily shortly after this was reported it was fixed.

However, I think that it would be nice to have some sort of configuration option where you could set the version number of any of the dependencies if you choose to. Not that I expect this sort of feature to be needed all the time, but, no one is perfect.

### Operating System

Linux, Windows

### FastAPI Version

0.1000

### Python Version

3.11 from Docker image (python:3.11-slim)

**Comment by YuriiMotov** [comment, marked as answer]
You can specify versions of dependencies and sub-dependencies in `pyproj.toml` file of your project: https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#dependencies-and-requirements

See also: [poetry](https://python-poetry.org/), [uv](https://docs.astral.sh/uv/)
